appinventor/appengine/src/com/google/appinventor/client/explorer/folder/ProjectFolder.java

```diff
@@ -0,0 +1,378 @@
+// Copyright 2021 MIT, All rights reserved
+// Released under the Apache License, Version 2.0
+// http://www.apache.org/licenses/LICENSE-2.0
+
+package com.google.appinventor.client.explorer.folder;
+
+import com.google.appinventor.client.Ode;
+import com.google.appinventor.client.OdeMessages;
+import com.google.appinventor.client.components.Icon;
+import com.google.appinventor.client.explorer.project.Project;
+import com.google.appinventor.client.explorer.youngandroid.ProjectListItem;
+import com.google.appinventor.client.explorer.project.ProjectSelectionChangeHandler;
+import com.google.gwt.core.client.GWT;
+import com.google.gwt.event.dom.client.ClickEvent;
+import com.google.gwt.i18n.client.DateTimeFormat;
+import com.google.gwt.json.client.JSONArray;
+import com.google.gwt.json.client.JSONObject;
+import com.google.gwt.json.client.JSONString;
+import com.google.gwt.uibinder.client.UiBinder;
+import com.google.gwt.uibinder.client.UiFactory;
+import com.google.gwt.uibinder.client.UiField;
+import com.google.gwt.uibinder.client.UiHandler;
+import com.google.gwt.user.client.ui.CheckBox;
+import com.google.gwt.user.client.ui.Composite;
+import com.google.gwt.user.client.ui.FlowPanel;
+import com.google.gwt.user.client.ui.Label;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import static com.google.appinventor.client.Ode.MESSAGES;
+import java.util.logging.Logger;
+
+
+public class ProjectFolder extends Composite {
+  private static final Logger LOG = Logger.getLogger(ProjectFolder.class.getName());
+  /**
+   * This class represents a folder containing project objects.
+   */
+  private String name;
+  private List<Project> projects = new ArrayList<>();
+  private List<ProjectListItem> projectListItems = new ArrayList<>();
+  private Map<String, ProjectFolder> folders = new HashMap<>();
+  private long dateCreated;
+  protected long dateModified;
+  private ProjectFolder parent;
+  protected JSONObject cachedJSON;
+  private ProjectSelectionChangeHandler changeHandler;
+
+  interface ProjectFolderUiBinder extends UiBinder<FlowPanel, ProjectFolder> {
+  }
+
+  private static final ProjectFolder.ProjectFolderUiBinder UI_BINDER = GWT.create(ProjectFolder.ProjectFolderUiBinder.class);
+  private boolean isExpanded = false;
+
+
+  @UiField
+  FlowPanel container;
+  @UiField
+  FlowPanel childrenContainer;
+  @UiField
+  Label nameLabel;
+  @UiField
+  Label dateModifiedLabel;
+  @UiField
+  Label dateCreatedLabel;
+  @UiField
+  CheckBox checkBox;
+  @UiField
+  Icon expandButton;
+
+  public ProjectFolder(String name, long dateCreated, long dateModified, ProjectFolder parent) {
+    initWidget(UI_BINDER.createAndBindUi(this));
+    this.name = name;
+    nameLabel.setText(name);
+    this.dateCreated = dateCreated;
+    this.dateCreatedLabel.setText(DateTimeFormat.getFormat(DateTimeFormat.PredefinedFormat.DATE_MEDIUM)
```

Comment:
A minor performance improvement would be caching the DateTimeFormat as a final static field and reusing it rather than getting it multiple times per folder.